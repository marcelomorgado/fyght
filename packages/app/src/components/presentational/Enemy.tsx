import React from "react";
import { Card, Alert } from "antd";
import { ChallengeModal } from "./ChallengeModal";
import { SkinAvatar } from "./SkinAvatar";
import { AvatarSize } from "../../constants";
import { formatDai, formatPercent } from "../../helpers";
import { BigNumber } from "ethers";
import { FyghterBalance } from "./FyghterBalance";

type Props = {
  myFyghter: Fyghter;
  enemy: Enemy;
};

// TODO: Move to .env
const BET_VALUE = `${5e18}`;
const ETHER = `${1e18}`;

const GainOrLoss: React.FC<{ value: BigNumber }> = ({ value }: { value: BigNumber }) => {
  const color = value.lt(BigNumber.from(0)) ? "red" : "blue";
  return <span style={{ color }}>{`${formatDai(value)}`}</span>;
};

// TODO: Compare with gainIfWin
const WinProbability: React.FC<{ value: BigNumber }> = ({ value }: { value: BigNumber }) => {
  const color = value.lt(BigNumber.from(`${5e17}`)) ? "red" : "blue";
  return <span style={{ color }}>{`${formatPercent(value)}`}</span>;
};

export const Enemy: React.FC<Props> = ({
  myFyghter,
  enemy: {
    fyghter: { id, skin, name, xp, balance },
    winProbability,
  },
}: Props) => {
  const probability = winProbability ? winProbability : BigNumber.from("0");
  const ifWin = BigNumber.from(BET_VALUE)
    .mul(BigNumber.from(ETHER).sub(probability))
    .div(ETHER);
  const ifLose = BigNumber.from(BET_VALUE)
    .mul(probability)
    .div(BigNumber.from(ETHER))
    .mul(BigNumber.from(-1));

  const enemyHasEnoughBalance = balance.gte(ifWin);

  let warningMessage = null;
  if (!myFyghter) {
    warningMessage = "You have to create your fyghter to be able to do challenges.";
  } else if (!myFyghter.balance.gte(ifLose)) {
    warningMessage = "Your fyghter with insufficient funds.";
  } else if (!enemyHasEnoughBalance) {
    warningMessage = "Enemy with insufficient funds.";
  }

  return (
    <Card type="inner" title={name} hoverable={true}>
      <SkinAvatar skin={skin} size={AvatarSize.MEDIUM} />
      <p></p>
      <p>{`XP: ${xp}`}</p>
      <p>
        {`Balance: `}
        <FyghterBalance value={balance} />
      </p>

      {!myFyghter ? (
        <></>
      ) : (
        <>
          <p>
            {`Win probability: `}
            <WinProbability value={probability} />
          </p>
          <p>
            {`Gain if win: `}
            <GainOrLoss value={ifWin} />
          </p>
          <p>
            {`Loss if lose: `}
            <GainOrLoss value={ifLose} />
          </p>
        </>
      )}

      <ChallengeModal enemyId={id} disabled={!!warningMessage} />
      {warningMessage ? (
        <>
          <p></p>
          <Alert message={warningMessage} type="warning" />
        </>
      ) : (
        <></>
      )}
    </Card>
  );
};
