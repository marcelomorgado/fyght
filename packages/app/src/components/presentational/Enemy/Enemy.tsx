import React from "react";
import { Card, Alert } from "antd";
import { ChallengeModal } from "../ChallengeModal";
import { SkinAvatar } from "../SkinAvatar";
import { AvatarSize } from "../../../constants";
import { formatDai, formatPercent, calculateGainAndLoss } from "../../../helpers";
import { BigNumber } from "ethers";
import { EnemyBalance } from "./EnemyBalance";

type Props = {
  myFyghter: Fyghter;
  enemy: Enemy;
};

const GainOrLoss: React.FC<{ value: BigNumber }> = ({ value }: { value: BigNumber }) => {
  const color = value.lt(BigNumber.from(0)) ? "red" : "blue";
  return <span style={{ color }}>{`${formatDai(value)}`}</span>;
};

const WinProbability: React.FC<{ value: BigNumber }> = ({ value }: { value: BigNumber }) => {
  const color = value.lt(BigNumber.from(`${5e17}`)) ? "red" : "blue";
  return <span style={{ color }}>{`${formatPercent(value)}`}</span>;
};

export const Enemy: React.FC<Props> = ({
  myFyghter,
  enemy: {
    fyghter: { id, skin, name, xp, balance },
    winProbability,
  },
}: Props) => {
  const { gainIfWin, lossIfLose } = calculateGainAndLoss(winProbability);

  const enemyHasEnoughBalance = balance.gte(gainIfWin);

  let warningMessage = null;
  if (!myFyghter) {
    warningMessage = "You have to create your fyghter to be able to do challenges.";
  } else if (!myFyghter.balance.gte(lossIfLose)) {
    warningMessage = "Your fyghter with insufficient funds.";
  } else if (!enemyHasEnoughBalance) {
    warningMessage = "Enemy with insufficient funds.";
  }

  return (
    <Card type="inner" title={name} hoverable={true}>
      <SkinAvatar skin={skin} size={AvatarSize.MEDIUM} />
      <p></p>
      <p>{`XP: ${xp}`}</p>
      <p>
        {`Balance: `}
        <EnemyBalance value={balance} gainIfWin={gainIfWin} />
      </p>

      {!myFyghter ? (
        <></>
      ) : (
        <>
          <p>
            {`Win probability: `}
            <WinProbability value={winProbability || BigNumber.from("0")} />
          </p>
          <p>
            {`Gain if win: `}
            <GainOrLoss value={gainIfWin} />
          </p>
          <p>
            {`Loss if lose: `}
            <GainOrLoss value={lossIfLose} />
          </p>
        </>
      )}

      <ChallengeModal enemyId={id} disabled={!!warningMessage} />
      {warningMessage ? (
        <>
          <p></p>
          <Alert message={warningMessage} type="warning" />
        </>
      ) : (
        <></>
      )}
    </Card>
  );
};
