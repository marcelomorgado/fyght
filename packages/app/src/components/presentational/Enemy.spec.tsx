import * as React from "react";
import { shallow } from "enzyme";

import { Enemy } from "./Enemy";

describe("Enemy", () => {
  test("should render the component", () => {
    expect(
      shallow(
        <Enemy
          enemy={{
            id: 1,
            name: "Charlie",
            skin: "normal_guy",
            xp: 2,
            qi: 3,
            winCount: 4,
            lossCount: 5,
          }}
        />
      )
    ).toMatchSnapshot();
  });
});
