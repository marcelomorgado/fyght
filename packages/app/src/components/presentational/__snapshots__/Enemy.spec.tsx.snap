// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Enemy should render the component 1`] = `
<Card
  hoverable={true}
  title="Charlie"
  type="inner"
>
  <Component
    size="medium"
    skin="normal_guy"
  />
  <p />
  <p>
    wins: 4 / losses: 5
  </p>
  <p>
    XP: 2
  </p>
  <AttackModal />
</Card>
`;
